Rust/src/random_cut_forest.rs

```diff
@@ -90,6 +92,60 @@ impl<T> RandomCutForest<T>
         }
     }
 
+    /// Returns the anomaly score associated with the input point relative to
+    /// the data used to update the random cut forest model.
+    ///
+    /// Each constituent random cut tree reports an anomaly score based on its
+    /// random sample of observed data. The forest anomaly score is the mean
+    /// of these per-tree scores.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{RandomCutForest, RandomCutForestBuilder};
+    ///
+    /// // create a default RCF on two-dimensional data
+    /// let mut rcf: RandomCutForest<f32> = RandomCutForestBuilder::new(2).build();
+    ///
+    /// // train the forest on some vector data
+    /// # let data: Vec<Vec<f32>> = vec![vec![0.0, 0.0], vec![1.0, 1.0]];
+    /// for point in data.iter() {
+    ///     rcf.update(point.clone());
+    /// }
+    ///
+    /// // compute an anomaly score on some query point
+    /// let query = vec![0.5, 0.5];
+    /// let score = rcf.anomaly_score(&query);
+    ///
+    /// // compute the anomaly scores of the training points
+    /// let scores: Vec<f32> = data.iter().map(|p| rcf.anomaly_score(p)).collect();
+    /// ```
+    pub fn anomaly_score(&self, point: &Vec<T>) -> T {
+        let anomaly_score: T = self.trees.iter()
+            .map(|t| self.tree_anomaly_score(point, t.tree()))
+            .sum();
+        anomaly_score / T::from(self.num_trees()).unwrap()
+    }
+
+    /// Compute the anomaly score from a particular tree. Used by
+    /// ['Self::anomaly_score'].
+    fn tree_anomaly_score(&self, point: &Vec<T>, tree: &Tree<T>) -> T {
+        // traverse nodes from the root to the leaf closest to the input point
+        let nodes: Vec<&Node<T>> = tree.traverse(point).collect();
+        let mut visitor = AnomalyScoreVisitor::new(tree, point);
```

Comment:
Instead of having an AnomalyScoreVisitor, couldn't the forest have an AnomalyScorer that implements a `score` function? This way, the scoring functionality RCF uses could be configurable at construction.

Reply:
Good thinking about future scoring functions. (As there are several known scoring functions and many many more unknown ones.) I'm not sure how best to go about this in Rust. Feel tree to cut a tracking ticket where we can start a discussion. Would be good to get people's ideas.

Initial idea: `Visitor` trait with [associated output type](https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html). `RandomCutForest` would have a single generic scoring function that accepts a point and a type that implements `Visitor`. To implement `anomaly_score()` you just create an instance of the visitor and give it to this generic scoring function. I dunno. Again, best to start an issue thread.

Reply:
#206